### DISCUSSION TITLE
Multiple issues / undocumented behaviors

### DISCUSSION THREAD

**Discussion by YellowFish085** [spatie/laravel-model-states#186] [Bugs]
Hello,

First thank you guys for all the packages you've created and maintained for Laravel, I couldn't imagine doing a Laravel project without at least a few of your packages in it.

I've been using this package for a few days and I've come across 2 issues. I do not know if these are actual issues or intended behaviors so I'll tag this discussion in the bugs category.

1\) In the `StateConfig` class, the [createTransitionKey()](https://github.com/spatie/laravel-model-states/blob/master/src/StateConfig.php#L104) method generates keys using the `"{$from}-{$to}"` format, and the [transitionableStates()](https://github.com/spatie/laravel-model-states/blob/master/src/StateConfig.php#L81) method splits the keys assuming that the key is indeed splittable in two with everything before `-` being the _from_ state and everything after  being the _to_ state.

Because of that, it's not possible to set the `$name` property of a state to a string with `-` in it because the  [transitionableStates()](https://github.com/spatie/laravel-model-states/blob/master/src/StateConfig.php#L81) will break :

```php
class StateA extends MyState
{
    // Invalid value, transitionableStates() will not be able to process transitions from and to this state.
    public static $name = 'state-a';

    // Valid value.
    public static $name = 'state_a';
}
```

This particularity is not documented in the [related documentation page](https://spatie.be/docs/laravel-model-states/v2/working-with-states/02-serializing-states), if this is an intended behavior or a side effect of how the package works it should be documented.

2\) I'm not 100% sure about the reason behind this issue, but I think that due to how the `State` class [resolves state classes](https://github.com/spatie/laravel-model-states/blob/master/src/State.php#L271), it's not possible to have the concrete state classes in a different directory than the abstract class.

As an example, using the classes [in the documentation](https://spatie.be/docs/laravel-model-states/v2/working-with-states/01-configuring-states), you must have the `Pending`, `Paid` and `Failed` classes in the same directory as the abstract `PaymentState` class they extend, else it won't work.

Once again, I do not know if this is a side effect of how the package works but it's also not documented in the package documentation.

That's all, thanks again for the work you're doing on all these packages and have a nice day!

**Comment by brendt** [comment]
1) that's a valid issue, I never considered that people would want to use `-` in their state names. We could change that token, feel free to submit a PR.
2) Auto discovered states must indeed live in the same directory as the abstract class. That's intended behaviour.
